### PR TITLE
Remove dependency on Sigil 4.7.0 / Sigil-vNext 4.8.41 (updated)

### PR DESCRIPTION
**Pull request by zacharycmontoya**
### Summary
This PR removes usages of `Sigil 4.7.0` and `Sigil-vNext 4.8.41`, removing `Sigil.dll` dependency on all target frameworks.

### Changes
- Remove `Sigil` NuGet packages from `Datadog.Trace.ClrProfiler.Managed.csproj` and add direct NuGet package references to `System.Reflection.Emit` NuGet packages that were previously transitively referenced for the `netstandard2.0` build target
- `Datadog.Trace.ClrProfiler.Emit.ObjectExtensions`
  - Replace all usages of `Sigil.Emit<T>` with `System.Reflection.Emit.ILGenerator`
  - Store static field `System.Reflection.Emit.ModuleBuilder Module` to hold single module in which `MethodBuilder` dynamic methods will be emitted
- `Datadog.Trace.ClrProfiler.Emit.MethodBuilder`
  - Replace all usages of `Sigil.Emit<T>` with `System.Reflection.Emit.ILGenerator`
- `Datadog.Trace.ClrProfiler.Emit.DynamicMethodBuilder'1`
  - Replace all usages of `Sigil.Emit<T>` with `System.Reflection.Emit.ILGenerator`
- Fix `ExecuteReaderAsync` in the `SqlCommandIntegration` and `NpgsqlCommandIntegration` classes that began failing after removing Sigil. The return types were specified by the delegate to be `Task<DbDataReader>` even though the public interfaces were specified to return `Task<SqlDataReader>` and `Task<NpqgsqlDataReader>`

### Benchmarks
```
|                               Method |     Toolchain |     Mean |     Error |    StdDev | Ratio | RatioSD |
|------------------------------------- |-------------- |---------:|----------:|----------:|------:|--------:|
|                   GetField_Old_Sigil |        net472 | 4.911 ns | 0.0865 ns | 0.0676 ns |  1.48 |    0.02 |
|                  GetField_New_ILEmit |        net472 | 4.810 ns | 0.0530 ns | 0.0414 ns |  1.36 |    0.01 |
|                                      |               |          |           |           |       |         |
|                   GetField_Old_Sigil | netcoreapp3.1 | 3.310 ns | 0.0164 ns | 0.0154 ns |  1.00 |    0.00 |
|                  GetField_New_ILEmit | netcoreapp3.1 | 3.549 ns | 0.0373 ns | 0.0291 ns |  1.00 |    0.00 |
|                                      |               |          |           |           |       |         |
|   CreateMethodCallDelegate_Old_Sigil |        net472 | 5.643 ns | 0.1426 ns | 0.1191 ns |  1.29 |    0.03 |
|  CreateMethodCallDelegate_New_ILEmit |        net472 | 4.619 ns | 0.1156 ns | 0.1331 ns |  1.38 |    0.04 |
|                                      |               |          |           |           |       |         |
|   CreateMethodCallDelegate_Old_Sigil | netcoreapp3.1 | 4.375 ns | 0.0439 ns | 0.0411 ns |  1.00 |    0.00 |
|  CreateMethodCallDelegate_New_ILEmit | netcoreapp3.1 | 3.367 ns | 0.0713 ns | 0.0595 ns |  1.00 |    0.00 |
|                                      |               |          |           |           |       |         |
|  MethodBuilderEmitDelegate_Old_Sigil |        net472 | 3.382 ns | 0.0280 ns | 0.0219 ns |  0.97 |    0.01 |
| MethodBuilderEmitDelegate_New_ILEmit |        net472 | 2.212 ns | 0.0746 ns | 0.0733 ns |  1.09 |    0.04 |
|                                      |               |          |           |           |       |         |
|  MethodBuilderEmitDelegate_Old_Sigil | netcoreapp3.1 | 3.486 ns | 0.0304 ns | 0.0269 ns |  1.00 |    0.00 |
| MethodBuilderEmitDelegate_New_ILEmit | netcoreapp3.1 | 2.023 ns | 0.0114 ns | 0.0089 ns |  1.00 |    0.00 |
```
@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/SqlCommandIntegration.cs [hunk 1]
```diff
@@ -218,11 +219,38 @@ public static class SqlCommandIntegration
                 throw;
             }
 
+            return AsyncHelper.InvokeGenericTaskDelegate(
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNet/SqlCommandIntegration.cs:222]: Kevin noted in the previous PR:

> You could instead do like HttpMessageHandlerIntegration:

> - Make the invoked async method return Task
> - Invoke it directly and use the TaskExtensions.Cast helper to convert the return type
> - In the invoked async method, cast the return value of instrumentedMethod to Task to await it, then use `.GetProperty("Result").Value` to fetch the result
If was faster for HttpMessageHandlerIntegration, so I expect the same here. But we can do that in another PR if you prefer

I'd prefer to keep this as-is for now in this PR.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs [hunk 2]
```diff
@@ -196,15 +196,15 @@ internal static class DynamicMethodBuilder<TDelegate>
 
             if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:197]: nit: instead comparing `returnType == typeof(object)` we should compare if returnType is ValueType or not.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs [hunk 3]
```diff
@@ -158,36 +157,37 @@ internal static class DynamicMethodBuilder<TDelegate>
...
+            ILGenerator il = dynamicMethod.GetILGenerator();
 
             // load each argument and cast or unbox as necessary
             for (ushort argumentIndex = 0; argumentIndex < parameterTypes.Length; argumentIndex++)
             {
                 Type delegateParameterType = parameterTypes[argumentIndex];
                 Type underlyingParameterType = effectiveParameterTypes[argumentIndex];
 
-                dynamicMethod.LoadArgument(argumentIndex);
+                il.Emit(OpCodes.Ldarg, argumentIndex);
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:169]: this implementation misses the shortcuts from `Ldarg_0` to `Ldarg_3` and the short version `Ldarg_S`. I don't know if there is a performance impact by using the int16 version directly.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs [hunk 4]
```diff
@@ -196,15 +196,15 @@ internal static class DynamicMethodBuilder<TDelegate>
 
             if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
             {
-                dynamicMethod.Box(methodInfo.ReturnType);
+                il.Emit(OpCodes.Box, methodInfo.ReturnType);
             }
             else if (methodInfo.ReturnType != returnType)
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:201]: If `returnType` is a value type, the `Castclass` is invalid.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/ObjectExtensions.cs [hunk 5]
```diff
@@ -231,31 +245,33 @@ private static PropertyFetcherCacheKey GetKey<TResult>(string name, Type type)
...
-            dynamicMethod.LoadArgument(0);
+            DynamicMethod dynamicMethod = new DynamicMethod($"{containerType.FullName}.{fieldName}", typeof(TResult), new Type[] { typeof(object) }, ObjectExtensions.Module, skipVisibility: true);
+            ILGenerator il = dynamicMethod.GetILGenerator();
+
+            il.Emit(OpCodes.Ldarg_0);
 
             if (containerType.IsValueType)
             {
-                dynamicMethod.UnboxAny(containerType);
+                il.Emit(OpCodes.Unbox, containerType);
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/ObjectExtensions.cs:255]: the conversion should be `Unbox_Any`

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs [hunk 6]
```diff
@@ -313,36 +313,37 @@ private TDelegate EmitDelegate()
...
+            ILGenerator il = dynamicMethod.GetILGenerator();
 
             // load each argument and cast or unbox as necessary
             for (ushort argumentIndex = 0; argumentIndex < delegateParameterTypes.Length; argumentIndex++)
             {
                 Type delegateParameterType = delegateParameterTypes[argumentIndex];
                 Type underlyingParameterType = effectiveParameterTypes[argumentIndex];
 
-                dynamicMethod.LoadArgument(argumentIndex);
+                il.Emit(OpCodes.Ldarg, argumentIndex);
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs:325]: this implementation misses the shortcuts from Ldarg_0 to Ldarg_3 and the short version Ldarg_S. I don't know if there is a performance impact by using the int16 version directly.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs [hunk 7]
```diff
@@ -351,15 +352,15 @@ private TDelegate EmitDelegate()
 
             if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs:353]: nit: instead comparing `returnType == typeof(object)` we should compare if returnType is ValueType or not.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs [hunk 8]
```diff
@@ -351,15 +352,15 @@ private TDelegate EmitDelegate()
...
             {
-                dynamicMethod.Box(methodInfo.ReturnType);
+                il.Emit(OpCodes.Box, methodInfo.ReturnType);
             }
             else if (methodInfo.ReturnType != returnType)
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs:357]: If `returnType` is a value type, the `Castclass` is invalid.

**Review by tonyredondo — CHANGES_REQUESTED** [review]
Overall looks good, I just left a couple of comments.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/ObjectExtensions.cs (hunk 5 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/ObjectExtensions.cs:255]: Thanks for spotting this! Fixed in 0ce85d3fe2582b68df47c0e74f64d295476ccf14

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 3 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:169]: Thanks for spotting this! In 0ce85d3fe2582b68df47c0e74f64d295476ccf14 I put together a small switch statement. Is that fine with you?

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 2 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:197]: I don't really want to change behavior right now, but this is some behavior that I think is fragile and requires us to specify return types correctly, which we have so far.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs (hunk 7 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs:353]: I don't really want to change behavior right now, but this is some behavior that I think is fragile and requires us to specify return types correctly, which we have so far.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 4 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:201]: You're right. If I use your suggestion to check if the returnType is a value type instead of just strictly equal to `typeof(object)`, I have the following code, but I don't really know what we should be doing in this additional case. Throw an exception since the method was incorrectly generated by us?

```csharp
            if (methodInfo.ReturnType.IsValueType && !returnType.IsValueType) // Changed from returnType == typeof(object)
            {
                il.Emit(OpCodes.Box, methodInfo.ReturnType);
            }
            else if (methodInfo.ReturnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                // Figure out what to do
            }
            else if (!methodInfo.ReturnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                il.Emit(OpCodes.Castclass, returnType);
            }
```

Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:201]: Yeah, I will expect an exception on those cases. (Internally we can just return without converting between two value types. That way is how Unsafe.As works but is dangerous to do that without being explicit)

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 3 above)
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:169]: I can't see 0ce85d3fe2582b68df47c0e74f64d295476ccf14 but yeah a switch case is the right thing to do here 😊

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs]: Oof, sorry about that. I just pushed my branch up

**Review by tonyredondo — APPROVED** [review]
LGTM 👌 .
The only thing is the fragile comparison to apply the `CastClass` but I guess is not part of this PR.

**Comment by zacharycmontoya** [comment]
@tonyredondo I will include the `Castclass` change in this PR so I'll be pushing another commit soon

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 4 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:218]: Another case I hadn't considered: If the target method is an object but the outer delegate is a value type. Should we attempt an unbox and if it's incompatible then that operation will throw an InvalidClassException, like so:

```csharp
            if (methodInfo.ReturnType.IsValueType && !returnType.IsValueType)
            {
                il.Emit(OpCodes.Box, methodInfo.ReturnType);
            }
            else if (methodInfo.ReturnType.IsValueType && returnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                throw new ArgumentException($"Cannot convert the target method's return type {methodInfo.ReturnType.FullName} (valuetype) to the delegate method's return type {returnType.FullName} (valuetype)");
            }
            else if (!methodInfo.ReturnType.IsValueType && returnType.IsValueType)
            {
                il.Emit(OpCodes.Unbox, returnType);
            }
            else if (!methodInfo.ReturnType.IsValueType && !returnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                il.Emit(OpCodes.Castclass, returnType);
            }
```

File: src/Datadog.Trace.DuckTyping/Datadog.Trace.DuckTyping.csproj [hunk 9]
```diff
@@ -19,7 +19,7 @@
   </PropertyGroup>
 
   <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0'">
-    <PackageReference Include="System.Reflection.Emit" Version="4.7.0" />
-    <PackageReference Include="System.Reflection.Emit.Lightweight" Version="4.7.0" />
+    <PackageReference Include="System.Reflection.Emit" Version="4.3.0" />
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.DuckTyping/Datadog.Trace.DuckTyping.csproj:22]: @tonyredondo FYI I brought the NuGet package reference down since this still has the API surface area

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs [hunk 10]
```diff
@@ -313,53 +313,79 @@ private TDelegate EmitDelegate()
...
-                dynamicMethod.CallVirtual(methodInfo);
+                il.Emit(OpCodes.Callvirt, methodInfo);
             }
             else
             {
                 throw new NotSupportedException($"OpCode {_originalOpCodeValue} not supported when calling a method.");
             }
 
-            if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
+            if (methodInfo.ReturnType.IsValueType && !returnType.IsValueType)
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs:370]: @tonyredondo I added more thorough error handling on the return types, please review

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs [hunk 11]
```diff
@@ -158,53 +157,79 @@ internal static class DynamicMethodBuilder<TDelegate>
...
-                dynamicMethod.CallVirtual(methodInfo);
+                il.Emit(OpCodes.Callvirt, methodInfo);
             }
             else
             {
                 throw new NotSupportedException($"OpCode {callOpCode} not supported when calling a method.");
             }
 
-            if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
+            if (methodInfo.ReturnType.IsValueType && !returnType.IsValueType)
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:214]: @tonyredondo I added more thorough error handling on the return types, please review

File: src/Datadog.Trace.DuckTyping/Datadog.Trace.DuckTyping.csproj (hunk 9 above)
Comment (tonyredondo) [reply src/Datadog.Trace.DuckTyping/Datadog.Trace.DuckTyping.csproj:22]: Oh yeah, makes sense.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 4 above)
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs]:
```
if (!methodInfo.ReturnType.IsValueType && returnType.IsValueType)
```
IMO, I prefer to throw an exception and not handle it as a best effort. As developer we have control of the `returnType`. Is an edge case a scenario where the target method returns an object and we want a value type. If the target method returns a class, we should change the returnType to that class or an object in case we don't have the class at compile time.

```
 if (methodInfo.ReturnType.IsValueType && returnType.IsValueType && methodInfo.ReturnType != returnType)
```
Now I'm thinking in this scenario where the target method returns an struct that we don't have at compile time. Actually what we can do is to create our own struct with the same layout at the target and just return without any conversion. I don't know if we ever gonna have that scenario, but at least there is a way to handle it in the future if we want to.

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs (hunk 10 above)
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs:370]: looks great :shipit:

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 11 above)
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs:214]: looks great :shipit:

File: src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs (hunk 4 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs]: The latter scenario with different value types is really interesting. If we have that scenario in the future, I think we can pick that up then

Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs]: 👍

**Review by lucaspimentel — APPROVED** [review]
